### ISSUE TITLE
Review converged configuration APIs

### ISSUE THREAD

**Issue by terrajobst** [dotnet/corefx#2199]
ASP.NET has created configuration APIs (TBD: Link). @divega was driving a working group to push these APIs down so that they can become general purpose BCL-ish APIs.

We should review the proposal for a converged API set.

**Comment by terrajobst** [comment]
We've reviewed the proposal and think it's not quite ready yet. Notes are available [here](https://github.com/dotnet/apireviews/tree/master/2015-06-30-configuration).

**Comment by Garciat** [comment]
(Copy of https://github.com/dotnet/apireviews/pull/6#issuecomment-117510482)

Hello! I have couple of ideas I'd like to share:

**One**
I gathered from the code review video that you were trying to model this (F#):

```fsharp
type Configuration =
    | KeyValue of string * string
    | Section of string * seq<Configuration>
```

ADT's can be tricky to implement in OO (especially recursive ones), as they mix each option's concerns into a single type.

Particularly, this equates to modeling Directory and File with a single class [as Java does](http://docs.oracle.com/javase/8/docs/api/java/io/File.html), which ends up having discriminator methods (`isDirectory()` and `isFile()`) and special case methods (`listFiles()` returns `null` if the path is not a directory). And then there's the problem with *mutability*. The path behind the `File` instance may change to point from an actual file to a directory and then suddenly the object behaves differently. Both these issues are present in the current Configuration module.

**Two**
Concept: use model binding for configuration.

*Rationale:*
One could argue configuration models are in fact static: components have fixed, well-defined configuration points. Treating configuration as a dictionary (or a tree in this case) of strings is a bit of an over-generalization.

*Pros:*
* Consumers no longer have to parse string values by hand.
* Type-safe, self-validating model.
* Could probably reuse model binding logic in `Microsoft.AspNet.Mvc.ModelBinding`.
* Tooling hints when editing configuration files thanks to data types.
* The library doesn't have to worry about the mutability of the user's configuration model.

*Cons:*
* Reflection-based.
* ...

*Samples:*
```csharp
// user-defined class with bits they care about
public class AppConfigurationModel
{
    public String BackgroundColor { get; set; }

    // let model binding parse values
    public Int32 SessionLifeMinutes { get; set; }

    // mutable, generic configuration.
    // using concurrent data structures, or not, is up to the user
    public ConcurrentDictionary<String, String> Etc { get; set; }
    
    // components may expose configuration model types for the user
    // which are then recursively bound just like in MVC
    public EntityFrameworkConfigurationModel Data { get; set; }
}

public class Startup
{
    public AppConfigurationModel Configuration { get; }
    
    public Startup(IApplicationEnvironment appEnv)
    {
        Configuration =
            new ConfigurationBuilder(appEnv.ApplicationBasePath)
            .AddJsonFile("config.json")
            .AddEnvironmentVariables()
            .Build<AppConfigurationModel>();
    }
    
    public void ConfigureServices(IServiceCollection services)
    {
        // feed to DI
        services.AddSingleton(appConfiguration);
        // could even do this and let EF pick up the config
        services.AddSingleton(appConfiguration.Data);
    }
}
```

```json
// config.json
{
  "backgroundColor": "blue",
  "sessionLifeMinutes": 100,
  "etc": {
    "hello": "world"
  },
  "data": {
    "defaultConnection": {
      "connectionString": "..."
    }
  }
}
```

We could even leverage attributes to select configuration sources:

```csharp
public class AppConfigurationModel
{
    [FromJson]
    public String DomainName { get; set; }

    [FromEnvironmentVariables("FOO_COUNT")]
    public Int32? NumberOfFoos { get; set; }
}
```

**Comment by RichiCoder1** [comment]
@Garciat I love the idea, but how does one push data back? I admit, I'm not overly familiar with the proposal, and whether or not it allows write as well as read. (Just saw this issue in my notifications ^_^)

**Comment by divega** [comment]
>  this equates to modeling Directory and File with a single class as Java does, which ends up having discriminator methods ( isDirectory()  and  isFile() )

@Garciat that is true but following the metaphor an IConfiguration can in fact represent both a "file"  and a "directory" at the same time.

**Comment by divega** [comment]
> use model binding for configuration

@Garciat 

We do have configuration binding. It is considered out of the scope of what we can bring to BCL because of the reflection requirement. Plus the API and implementation needs some polishing:

https://github.com/aspnet/Configuration/tree/dev/src/Microsoft.Framework.Configuration.Binder

**Comment by Garciat** [comment]
@divega I consider `IConfiguration.Get` and `IConfiguration.GetConfigurationSections` analogous to reading a file and listing a directory, respectively. Hence my observation of mixed responsibilities.

**Comment by Garciat** [comment]
@divega I saw no mention (or possibly missed it) of binding in the review video, so I assumed there was none. It's good to know otherwise.

Yeah, my proposal was not really aiming for BCL inclusion. I probably should've looked for a better PR/issue to comment on; my bad.
